contributor-docs/adrs/adr-019-responsive-values.md

```diff
@@ -0,0 +1,157 @@
+# ADR 019: Responsive Values
+
+## Status
+
+| Stage    | Status |
+| -------- | ------ |
+| Approved |        |
+| Adopted  |        |
+
+## Context
+
+There are situations where a component prop can support a value or a "responsive
+value" that results in the value changing depending on the viewport size. For
+example:
+
+```tsx
+// Value
+<Stack gap="condensed" />
+
+// Responsive value
+<Stack gap={{ narrow: 'condensed', regular: 'normal', wide: 'spacious' />
+```
+
+Typically, a responsive value will support all of our named breakpoints: narrow,
+regular, and wide. Authors can use a hook called `useResponsiveValue` to support
+resolving the value of a prop based on the current viewport size. The
+implementation of this hook uses [`matchMedia`](https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia)
+which, unfortunately, has a downside: the value of a prop may shift when the
+component is server-side rendered. If the prop is used for styles or layout, then this will lead to a layout shift
+when the component hydrates and the viewport size is different than the fallback
+size on the server.
+
+As a result, we would like to offer a paved path for providing responsive props
+in a way tha twill not lead to layout shift during server-side rendering.
+
+## Decision
+
+Authors should use data attributes to bridge between responsive values in
+JavaScript and the corresponding styles in CSS. Authors should avoid using
+JavaScript-driven mechanisms for controlling responsive values in order to
+prevent layout shift during SSR.
+
+For example, consider a `Stack` component that has a `gap` prop. This prop can
+receive a plain value or one that is a responsive value. To support this,
+authors should use the following patterns:
+
+```tsx
+type GapScale = 'none' | 'condensed' | 'normal' | 'spacious'
+
+type StackProps = React.PropsWithChildren<{
+  // Responsive props should be a union of the value of the prop plus
+  // `ResponsiveValue` over that value
+  gap?: GapScale | ResponsiveValue<GapScale>
+}>
+
+function Stack({children, gap}: StackProps) {
+  // The `getResponsiveAttributes` helper is useful for adding the corresponding
+  // `data-*` attributes to the container for the compnoent.
+  //
+  // In this case, it will add `data-gap` and `data-gap-{breakpoint}` attributes
+  // with the corresponding value set
+  return <StyledStack {...getResponsiveAttributes('gap', gap)}>{children}</StyledStack>
+}
+
+const StyledStack = styled.div`
+  &[data-gap='none'],
+  &[data-gap-narrow='none'] {
+    /* ... */
+  }
+
+  &[data-gap='condensed'],
+  &[data-gap-narrow='condensed'] {
+    /* ... */
+  }
+
+  /* ... */
+
+  // @custom-media --viewportRange-regular
+  @media (min-width: 768px) {
+    &[data-gap-regular='none'] {
+      /* ... */
+    }
+
+    &[data-gap-regular='condensed'] {
+      /* ... */
+    }
+
+    /* ... */
+  }
+
+  // @custom-media --viewportRange-wide
+  @media (min-width: 1400px) {
```

Comment:
@broccolinisoup great question, I'm actually not sure. I think it could be okay as it is piped to `sx` and could follow the same rules as defining this in `styled` directly. 

I think the only notable reason why we might prefer authoring in one format over the other is how easy it is to eventually transition to CSS Modules.

Reply:
That makes so much sense!